samza-core/src/main/java/org/apache/samza/operators/impl/store/TimeSeriesStoreImpl.java

```diff
@@ -0,0 +1,185 @@
+    this(kvStore, true);
+  }
+
+  @Override
+  public void put(K key, V val, Long timeStamp) {
+    // For append mode, values are differentiated by an unique sequence number. For overwrite mode, the sequence
+    // number is always zero. This ensures that only the most recent value is retained.
+    if (appendMode) {
+      incrementSeqNum();
+    }
+    TimeSeriesKey<K> timeSeriesKey = new TimeSeriesKey<>(key, timeStamp, seqNum);
+    kvStore.put(timeSeriesKey, val);
+  }
+
+  @Override
+  public ClosableIterator<TimeSeriesValue<V>> get(K key, Long startTimestamp, Long endTimeStamp) {
+    validateRanges(startTimestamp, endTimeStamp);
+    TimeSeriesKey<K> fromKey = new TimeSeriesKey(key, startTimestamp, 0);
+    TimeSeriesKey<K> toKey = new TimeSeriesKey(key, endTimeStamp, 0);
+
+    KeyValueIterator<TimeSeriesKey<K>, V> range = kvStore.range(fromKey, toKey);
+    return new TimeSeriesStoreIterator<>(range);
+  }
+
+  @Override
+  public void remove(K key, Long startTimestamp, Long endTimeStamp) {
+    validateRanges(startTimestamp, endTimeStamp);
+    TimeSeriesKey<K> fromKey = new TimeSeriesKey(key, startTimestamp, 0);
+    TimeSeriesKey<K> toKey = new TimeSeriesKey(key, endTimeStamp, 0);
+
+    List<TimeSeriesKey<K>> keysToDelete = new LinkedList<>();
+
+    KeyValueIterator<TimeSeriesKey<K>, V> range = kvStore.range(fromKey, toKey);
+    while (range.hasNext()) {
+      keysToDelete.add(range.next().getKey());
+    }
+
+    kvStore.deleteAll(keysToDelete);
+  }
+
+  @Override
+  public void close() {
+    kvStore.close();
+  }
+
+  private void incrementSeqNum() {
+    synchronized (seqNumLock) {
+      // cycles the seqNum back without a modulo operation
+      seqNum = (seqNum + 1) & Integer.MAX_VALUE;
+    }
+  }
+
+  private void validateRanges(Long startTimestamp, Long endTimeStamp) throws IllegalArgumentException {
+    if (startTimestamp < 0) {
+      throw new IllegalArgumentException(String.format("Start timestamp :%d is less than zero", startTimestamp));
+    }
+
+    if (endTimeStamp < 0) {
+      throw new IllegalArgumentException(String.format("End timestamp :%d is less than zero", endTimeStamp));
+    }
+
+    if (endTimeStamp < startTimestamp) {
+      throw new IllegalArgumentException(String.format("End timestamp :%d is less than start timestamp: %d", endTimeStamp, startTimestamp));
+    }
+  }
+
+  private static class TimeSeriesStoreIterator<K, V> implements ClosableIterator<TimeSeriesValue<V>> {
+
+    private final KeyValueIterator<TimeSeriesKey<K>, V> wrappedIterator;
+
+    public TimeSeriesStoreIterator(KeyValueIterator<TimeSeriesKey<K>, V> wrappedIterator) {
+      this.wrappedIterator = wrappedIterator;
+    }
+
+    @Override
+    public void close() {
+      wrappedIterator.close();
+    }
+
+    @Override
+    public boolean hasNext() {
+      return wrappedIterator.hasNext();
+    }
+
+    @Override
+    public TimeSeriesValue<V> next() {
```

Comment:
Should this return the key as well? See question about `range`.

Reply:
Since all access to the store are based on a key `K` and a timestamp, it's not necessary for the key to be returned again as a part of the iterator.